File: server/internal/usecase/lending.go
package usecase

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/kotarowakuwaku/asset-wish/server/internal/domain"
)

type LendingUsecase struct {
	tx           TxManager
	lendings     LendingRepository
	accounts     AccountRepository
	transactions TransactionRepository
	now          Clock
	newID        IDGenerator
}

func NewLendingUsecase(
	tx TxManager,
	lendings LendingRepository,
	accounts AccountRepository,
	transactions TransactionRepository,
	now Clock,
	newID IDGenerator,
) *LendingUsecase {
	return &LendingUsecase{
		tx:           tx,
		lendings:     lendings,
		accounts:     accounts,
		transactions: transactions,
		now:          now,
		newID:        newID,
	}
}

func (u *LendingUsecase) List(ctx context.Context, outstandingOnly bool) ([]domain.Lending, error) {
	return u.lendings.List(ctx, outstandingOnly)
}

// Create は立替を登録し、口座残高を減らして履歴を残す（detailed-design 3.2.1）。
//
// 立て替えた時点で自分の口座からは金が出ている。残高を減らさないと、
// 未回収額と残高の両方に同じ金を数えることになる。
func (u *LendingUsecase) Create(
	ctx context.Context,
	counterparty, description string,
	amount domain.Money,
	occurredOn time.Time,
	accountID uuid.UUID,
) (domain.Lending, error) {
	l, err := domain.NewLending(u.newID(), counterparty, description, amount, occurredOn)
	if err != nil {
		return domain.Lending{}, err
	}

	err = u.tx.RunInTx(ctx, func(ctx context.Context) error {
		account, err := u.accounts.Get(ctx, accountID)
		if err != nil {
			return err
		}
		account.ApplyDelta(-amount, u.now())
		if err := u.accounts.Update(ctx, account); err != nil {
			return err
		}
		if err := u.lendings.Create(ctx, l); err != nil {
			return err
		}
		return u.recordTransaction(ctx, accountID, -amount, domain.TransactionLendingCreated, l.ID, occurredOn)
	})
	if err != nil {
		return domain.Lending{}, err
	}
	return l, nil
}

// Collect は回収を記録し、口座残高を戻して履歴を残す（detailed-design 3.2.2）。
//
// 取得から更新までをトランザクションの内側に置く。取得と更新の間に別の
// 回収が入ると、どちらも未回収残高の範囲内に見えて過回収が成立しうるため
// （不変条件4）。
func (u *LendingUsecase) Collect(
	ctx context.Context,
	lendingID uuid.UUID,
	amount domain.Money,
	occurredOn time.Time,
	accountID uuid.UUID,
) (domain.Lending, error) {
	var collected domain.Lending

	err := u.tx.RunInTx(ctx, func(ctx context.Context) error {
		l, err := u.lendings.Get(ctx, lendingID)
		if err != nil {
			return err
		}
		if err := l.Collect(amount); err != nil {
			return err
		}
		if err := u.lendings.UpdateCollected(ctx, l); err != nil {
			return err
		}

		account, err := u.accounts.Get(ctx, accountID)
		if err != nil {
			return err
		}
		account.ApplyDelta(amount, u.now())
		if err := u.accounts.Update(ctx, account); err != nil {
			return err
		}

		if err := u.recordTransaction(
			ctx, accountID, amount, domain.TransactionLendingCollected, lendingID, occurredOn,
		); err != nil {
			return err
		}

		collected = l
		return nil
	})
	if err != nil {
		return domain.Lending{}, err
	}
	return collected, nil
}

func (u *LendingUsecase) Delete(ctx context.Context, id uuid.UUID) error {
	if _, err := u.lendings.Get(ctx, id); err != nil {
		return err
	}
	return u.lendings.Delete(ctx, id)
}

func (u *LendingUsecase) recordTransaction(
	ctx context.Context,
	accountID uuid.UUID,
	amount domain.Money,
	kind domain.TransactionKind,
	refID uuid.UUID,
	occurredOn time.Time,
) error {
	t, err := domain.NewTransaction(u.newID(), accountID, amount, kind, &refID, occurredOn)
	if err != nil {
		return err
	}
	return u.transactions.Create(ctx, t)
}

File: worker/src/usecase/lending.ts
import { Lending } from '../domain/lending'
import type { Money } from '../domain/money'
import type { IsoDate } from '../domain/time'
import type { AtomicWriter, IDGenerator, LendingRepository, WriteOperation } from './port'

/**
 * 立替の手順。
 *
 * **口座残高も取引履歴も触らない（不変条件4）。** 立て替えた時点で現金が
 * 出たとは限らないため（カード払いなら引き落としはまだ）、「立替 ＝ 現金が
 * 出た」という決め打ちをやめた。未回収額は実質資産の外の参考値として出す。
 *
 * 書き込みが1行だけになっても AtomicWriter を通し続けているのは、回収に
 * 楽観ロックが要るため。読み取り時の回収額を条件にしないと、同時に2回
 * 回収したときに両方が未回収残高の範囲内に見えて過回収が成立する。
 * その仕組みは AtomicWriter が持っている（不変条件10）。
 */
export class LendingUsecase {
  readonly #writer: AtomicWriter
  readonly #lendings: LendingRepository
  readonly #newID: IDGenerator

  constructor(writer: AtomicWriter, lendings: LendingRepository, newID: IDGenerator) {
    this.#writer = writer
    this.#lendings = lendings
    this.#newID = newID
  }

  list(outstandingOnly: boolean): Promise<Lending[]> {
    return this.#lendings.list(outstandingOnly)
  }

  /** 立替を登録する。口座残高は動かさない。 */
  async create(
    counterparty: string,
    description: string,
    amount: Money,
    occurredOn: IsoDate,
  ): Promise<Lending> {
    const lending = Lending.create(this.#newID(), counterparty, description, amount, occurredOn)
    await this.#writer.writeAll([{ kind: 'createLending', lending }])
    return lending
  }

  /**
   * 回収を記録する。入金先の口座は指定しない。
   *
   * 「読み取った時点から回収額が変わっていないこと」を条件に書き込む。取得と
   * 書き込みの間に別の回収が入ると、どちらも未回収残高の範囲内に見えて
   * 過回収が成立しうるため（不変条件4）。
   *
   * 回収日を受け取らないのは、記録する先が無くなったため。以前は取引履歴の
   * occurred_on に入れていたが、口座を触らなくなった今それは作られない。
   * 使わない引数を残すと「日付を渡したのに残らない」ことになる。
   */
  async collect(lendingId: string, amount: Money): Promise<Lending> {
    const lending = await this.#lendings.get(lendingId)
    const expectedCollectedAmount = lending.collectedAmount
    // 回収してよいかの判定は domain が持つ。usecase は結果を書くだけ。
    lending.collect(amount)

    const ops: WriteOperation[] = [
      { kind: 'updateLendingCollected', lending, expectedCollectedAmount },
    ]
    await this.#writer.writeAll(ops)
    return lending
  }

  /** 立替を削除する。無ければ NotFoundError。 */
  async delete(id: string): Promise<void> {
    await this.#lendings.get(id)
    await this.#lendings.delete(id)
  }
}
